controller/controller-common/src/main/java/com/pinterest/rocksplicator/controller/mysql/MySQLTaskQueue.java

```diff
@@ -17,71 +17,130 @@
+  public MySQLTaskQueue(EntityManager entityManager) {
+    this.entityManager = entityManager;
+  }
+
+  public boolean isClusterLocked(final String clusterName) throws MySQLTaskQueueException {
+    List<Integer> result = entityManager
+        .createNamedQuery("tag.findLock")
+        .setParameter("name",clusterName)
+        .getResultList();
+    if (result.isEmpty()) {
+      throw new MySQLTaskQueueException("Cluster doens't exist");
+    }
+    Integer locks = result.get(0);
+    return locks == 1;
   }
 
   @Override
   public boolean createCluster(final String clusterName) {
-    String insertQuery = String.format(
-        INSERT_TAG_QUERY_TEMPLATE, TAG_TABLE_NAME, clusterName, 0);
-    return JdbcUtils.executeUpdateQuery(connection, insertQuery);
+    TagEntity cluster = entityManager.find(TagEntity.class, clusterName);
+    if (cluster != null) {
+      LOG.error("Cluster {} is already existed", clusterName);
+      return false;
+    }
+    TagEntity tagEntity = new TagEntity().setName(clusterName);
+    entityManager.getTransaction().begin();
+    entityManager.persist(tagEntity);
+    entityManager.getTransaction().commit();
+    return true;
   }
 
   @Override
   public boolean lockCluster(final String clusterName) {
-    String lockQuery = String.format(LOCK_CLUSTER_QUERY_TEMPLATE, clusterName);
-    return JdbcUtils.executeUpdateQuery(connection, lockQuery);
+    boolean locked = true;
+    try {
+      locked = isClusterLocked(clusterName);
+    } catch (MySQLTaskQueueException e) {
+      LOG.error("Cluster {} hasn't been created", clusterName);
+    } finally {
+      if (locked) {
+        LOG.error("Cluster {} is already locked, cannot double lock", clusterName);
+        return false;
+      }
+    }
+    Query query = entityManager.createNamedQuery("tag.lock").setParameter("name", clusterName);
+    try {
+      entityManager.getTransaction().begin();
+      query.executeUpdate();
+      entityManager.getTransaction().commit();
+    } catch (IllegalStateException | PersistenceException e) {
+      LOG.error("Error locking cluster {}", clusterName, e);
+      return false;
+    }
+    return true;
   }
 
   @Override
   public boolean unlockCluster(final String clusterName) {
-    String unlockQuery = String.format(UNLOCK_CLUSTER_QUERY_TEMPLATE, clusterName);
-    return JdbcUtils.executeUpdateQuery(connection, unlockQuery);
+    Query query = entityManager.createNamedQuery("tag.unlock").setParameter("name", clusterName);
+    try {
+      entityManager.getTransaction().begin();
+      query.executeUpdate();
+      entityManager.getTransaction().commit();
+    } catch (IllegalStateException | PersistenceException e) {
+      LOG.error("Error unlocking cluster {}", clusterName, e);
+      return false;
+    }
+    return true;
+  }
+
+
+  @Override
+  public boolean removeCluster(final String clusterName) {
+    boolean locked = true;
+    try {
+      locked = isClusterLocked(clusterName);
```

Comment:
same thought as above, separate for precondition and actual transaction. (written in the doc)